File: chkit_python/src/chkit/core/sql.py
"""Render canonical schema definitions to ClickHouse DDL."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, TypeAlias

from pydantic import TypeAdapter

from chkit.core.codec import render_codec
from chkit.core.key_clause import is_plain_column_reference, normalize_key_columns
from chkit.core.model import (
    ColumnDefinition,
    DictionaryAttribute,
    DictionaryDefinition,
    MaterializedViewDefinition,
    MaterializedViewRefresh,
    ProjectionDefinition,
    SchemaDefinition,
    SkipIndexBloomFilter,
    SkipIndexDefinition,
    SkipIndexMinmax,
    SkipIndexSet,
    SkipIndexText,
    SkipIndexTokenBF,
    TableDefinition,
    TableRef,
    ViewDefinition,
)
from chkit.core.projection import render_projection_body
from chkit.core.text_index import render_text_index_type
from chkit.core.validate import assert_valid_definitions

_COLUMN_ADAPTER: TypeAdapter[ColumnDefinition] = TypeAdapter(ColumnDefinition)
_INDEX_ADAPTER: TypeAdapter[SkipIndexDefinition] = TypeAdapter(SkipIndexDefinition)
_PROJECTION_ADAPTER: TypeAdapter[ProjectionDefinition] = TypeAdapter(ProjectionDefinition)

ColumnInput: TypeAlias = ColumnDefinition | Mapping[str, Any]
IndexInput: TypeAlias = SkipIndexDefinition | Mapping[str, Any]
ProjectionInput: TypeAlias = ProjectionDefinition | Mapping[str, Any]


def _normalize_column(column: ColumnInput) -> ColumnDefinition:
    if isinstance(column, Mapping):
        return _COLUMN_ADAPTER.validate_python(dict(column))
    return column


def _normalize_index(index: IndexInput) -> SkipIndexDefinition:
    if isinstance(index, Mapping):
        return _INDEX_ADAPTER.validate_python(dict(index))
    return index


def _normalize_projection(projection: ProjectionInput) -> ProjectionDefinition:
    if isinstance(projection, Mapping):
        return _PROJECTION_ADAPTER.validate_python(dict(projection))
    return projection


def _render_default(value: str | int | float | bool) -> str:
    if isinstance(value, str):
        if value.startswith("fn:"):
            return value[3:]
        escaped = value.replace("'", "''")
        return f"'{escaped}'"
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def _render_column(col: ColumnDefinition) -> str:
    type_text = f"Nullable({col.type})" if col.nullable else f"{col.type}"
    out = f"`{col.name}` {type_text}"
    if col.default is not None:
        out += f" DEFAULT {_render_default(col.default)}"
    if col.comment is not None and len(col.comment) > 0:
        escaped = col.comment.replace("'", "''")
        out += f" COMMENT '{escaped}'"
    if col.codec is not None:
        out += f" {render_codec(col.codec)}"
    return out


def _render_key_clause_columns(columns: list[str], column_names: set[str]) -> str:
    # Quote a token when it names a declared column (including names that need
    # quoting like `user-id`) or is a bare identifier. Only true expressions
    # (e.g. `toStartOfHour(ts)`) are emitted verbatim.
    return ", ".join(
        f"`{c}`" if c in column_names or is_plain_column_reference(c) else c
        for c in normalize_key_columns(columns)
    )


def _render_index_type(idx: SkipIndexDefinition) -> str:
    if isinstance(idx, SkipIndexMinmax):
        return "minmax"
    if isinstance(idx, SkipIndexSet):
        return f"set({idx.max_rows})"
    if isinstance(idx, SkipIndexBloomFilter):
        rate = idx.false_positive_rate
        return "bloom_filter" if rate is None else f"bloom_filter({rate})"
    if isinstance(idx, SkipIndexTokenBF):
        return f"tokenbf_v1({idx.size_bytes}, {idx.hash_functions}, {idx.random_seed})"
    if isinstance(idx, SkipIndexText):
        return render_text_index_type(idx)
    # SkipIndexNgramBF is the only remaining variant in the discriminated union.
    return (
        f"ngrambf_v1({idx.ngram_size}, {idx.size_bytes}, {idx.hash_functions}, "
        f"{idx.random_seed})"
    )


def _render_setting_value(value: str | int | float | bool) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def _render_settings_clause(settings: dict[str, str | int | float | bool]) -> str:
    parts = [f"{k} = {_render_setting_value(v)}" for k, v in settings.items()]
    return ", ".join(parts)


def _render_projection(p: ProjectionDefinition) -> str:
    return f"PROJECTION `{p.name}` {render_projection_body(p)}"


def _render_index_line(idx: SkipIndexDefinition) -> str:
    return (
        f"INDEX `{idx.name}` ({idx.expression}) "
        f"TYPE {_render_index_type(idx)} GRANULARITY {idx.granularity}"
    )


def _render_table_sql(definition: TableDefinition) -> str:
    columns = [_render_column(c) for c in definition.columns]
    indexes_block = [_render_index_line(idx) for idx in (definition.indexes or [])]
    projections_block = [_render_projection(p) for p in (definition.projections or [])]
    body = ",\n  ".join(columns + indexes_block + projections_block)

    column_names = {column.name for column in definition.columns}
    clauses: list[str] = []
    if definition.partition_by is not None:
        clauses.append(f"PARTITION BY {definition.partition_by}")
    clauses.append(
        f"PRIMARY KEY ({_render_key_clause_columns(definition.primary_key, column_names)})"
    )
    clauses.append(
        f"ORDER BY ({_render_key_clause_columns(definition.order_by, column_names)})"
    )
    if definition.unique_key is not None and len(definition.unique_key) > 0:
        clauses.append(
            f"UNIQUE KEY ({_render_key_clause_columns(definition.unique_key, column_names)})"
        )
    if definition.ttl is not None:
        clauses.append(f"TTL {definition.ttl}")
    if definition.settings is not None and len(definition.settings) > 0:
        clauses.append(f"SETTINGS {_render_settings_clause(definition.settings)}")
    if definition.comment is not None and len(definition.comment) > 0:
        escaped = definition.comment.replace("'", "''")
        clauses.append(f"COMMENT '{escaped}'")

    return (
        f"CREATE TABLE IF NOT EXISTS {definition.database}.{definition.name}\n"
        f"(\n  {body}\n) ENGINE = {definition.engine}\n"
        f"{chr(10).join(clauses)};"
    )


def _render_view_sql(definition: ViewDefinition) -> str:
    return (
        f"CREATE VIEW IF NOT EXISTS {definition.database}.{definition.name} AS\n"
        f"{definition.as_};"
    )


def _render_refresh_settings(settings: dict[str, str | int | float]) -> str:
    parts: list[str] = []
    for k, v in settings.items():
        if isinstance(v, str):
            escaped = v.replace("'", "''")
            parts.append(f"{k} = '{escaped}'")
        else:
            parts.append(f"{k} = {v}")
    return ", ".join(parts)


def _render_depends_on(depends_on: list[TableRef]) -> str:
    return ", ".join(f"{d.database}.{d.name}" for d in depends_on)


def _render_refresh_clause(refresh: MaterializedViewRefresh) -> str:
    parts: list[str] = []
    if refresh.every is not None:
        parts.append(f"REFRESH EVERY {refresh.every}")
    elif refresh.after is not None:
        parts.append(f"REFRESH AFTER {refresh.after}")
    if refresh.offset is not None:
        parts.append(f"OFFSET {refresh.offset}")
    if refresh.randomize is not None:
        parts.append(f"RANDOMIZE FOR {refresh.randomize}")
    if refresh.depends_on is not None and len(refresh.depends_on) > 0:
        parts.append(f"DEPENDS ON {_render_depends_on(refresh.depends_on)}")
    if refresh.settings is not None and len(refresh.settings) > 0:
        parts.append(f"SETTINGS {_render_refresh_settings(refresh.settings)}")
    if refresh.append:
        parts.append("APPEND")
    return " ".join(parts)


def _render_materialized_view_sql(definition: MaterializedViewDefinition) -> str:
    header = (
        f"CREATE MATERIALIZED VIEW IF NOT EXISTS "
        f"{definition.database}.{definition.name}"
    )
    refresh_block = (
        f"\n{_render_refresh_clause(definition.refresh)}"
        if definition.refresh is not None
        else ""
    )
    to_clause = f" TO {definition.to.database}.{definition.to.name}"
    empty_clause = " EMPTY" if (definition.refresh is not None and definition.refresh.empty) else ""
    return (
        f"{header}{refresh_block}{to_clause}{empty_clause} AS\n{definition.as_};"
    )


def render_alter_modify_refresh(definition: MaterializedViewDefinition) -> str:
    if definition.refresh is None:
        msg = (
            f"Cannot render MODIFY REFRESH for "
            f"{definition.database}.{definition.name}: refresh is not set"
        )
        raise ValueError(msg)
    clause = _render_refresh_clause(definition.refresh)
    return f"ALTER TABLE {definition.database}.{definition.name} MODIFY {clause};"


def _render_dictionary_attribute(attr: DictionaryAttribute) -> str:
    out = f"`{attr.name}` {attr.type}"
    if attr.expression is not None:
        out += f" EXPRESSION {attr.expression}"
    elif attr.default is not None:
        out += f" DEFAULT {_render_default(attr.default)}"
    if attr.hierarchical:
        out += " HIERARCHICAL"
    if attr.bidirectional:
        out += " BIDIRECTIONAL"
    if attr.injective:
        out += " INJECTIVE"
    if attr.is_object_id:
        out += " IS_OBJECT_ID"
    return out


def _render_dictionary_range_column(column: str, column_names: set[str]) -> str:
    if column in column_names or is_plain_column_reference(column):
        return f"`{column}`"
    return column


def _render_dictionary_settings(settings: dict[str, str | int | float]) -> str:
    parts: list[str] = []
    for key, value in settings.items():
        if isinstance(value, str):
            escaped = value.replace("'", "''")
            parts.append(f"{key} = '{escaped}'")
        else:
            parts.append(f"{key} = {value}")
    return ", ".join(parts)


def render_dictionary_sql(definition: DictionaryDefinition, replace: bool = False) -> str:
    verb = (
        "CREATE OR REPLACE DICTIONARY" if replace else "CREATE DICTIONARY IF NOT EXISTS"
    )
    attrs = ",\n  ".join(
        _render_dictionary_attribute(a) for a in definition.attributes
    )
    column_names = {attribute.name for attribute in definition.attributes}
    pk = _render_key_clause_columns(definition.primary_key, column_names)
    clauses = [
        f"PRIMARY KEY {pk}",
        f"SOURCE({definition.source})",
        f"LAYOUT({definition.layout})",
        f"LIFETIME({definition.lifetime})",
    ]
    if definition.range is not None:
        min_col = _render_dictionary_range_column(definition.range.min, column_names)
        max_col = _render_dictionary_range_column(definition.range.max, column_names)
        clauses.append(f"RANGE(MIN {min_col} MAX {max_col})")
    if definition.settings is not None and len(definition.settings) > 0:
        clauses.append(f"SETTINGS({_render_dictionary_settings(definition.settings)})")
    if definition.comment:
        escaped = definition.comment.replace("'", "''")
        clauses.append(f"COMMENT '{escaped}'")
    return (
        f"{verb} {definition.database}.{definition.name}\n"
        f"(\n  {attrs}\n)\n{chr(10).join(clauses)};"
    )


def to_create_sql(definition: SchemaDefinition) -> str:
    assert_valid_definitions([definition])
    if isinstance(definition, TableDefinition):
        return _render_table_sql(definition)
    if isinstance(definition, ViewDefinition):
        return _render_view_sql(definition)
    if isinstance(definition, DictionaryDefinition):
        return render_dictionary_sql(definition)
    return _render_materialized_view_sql(definition)


def render_alter_add_column(definition: TableDefinition, column: ColumnInput) -> str:
    normalized = _normalize_column(column)
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"ADD COLUMN IF NOT EXISTS {_render_column(normalized)};"
    )


def render_alter_modify_column(definition: TableDefinition, column: ColumnInput) -> str:
    normalized = _normalize_column(column)
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"MODIFY COLUMN {_render_column(normalized)};"
    )


def render_alter_drop_column(definition: TableDefinition, column_name: str) -> str:
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"DROP COLUMN IF EXISTS `{column_name}`;"
    )


def render_alter_remove_codec(definition: TableDefinition, column_name: str) -> str:
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"MODIFY COLUMN `{column_name}` REMOVE CODEC;"
    )


def render_alter_add_index(definition: TableDefinition, index: IndexInput) -> str:
    normalized = _normalize_index(index)
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"ADD INDEX IF NOT EXISTS `{normalized.name}` ({normalized.expression}) "
        f"TYPE {_render_index_type(normalized)} GRANULARITY {normalized.granularity};"
    )


def render_alter_drop_index(definition: TableDefinition, index_name: str) -> str:
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"DROP INDEX IF EXISTS `{index_name}`;"
    )


def render_alter_add_projection(
    definition: TableDefinition, projection: ProjectionInput
) -> str:
    normalized = _normalize_projection(projection)
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"ADD PROJECTION IF NOT EXISTS `{normalized.name}` "
        f"{render_projection_body(normalized)};"
    )


def render_alter_drop_projection(definition: TableDefinition, projection_name: str) -> str:
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"DROP PROJECTION IF EXISTS `{projection_name}`;"
    )


def render_alter_modify_setting(
    definition: TableDefinition, key: str, value: str | int | float | bool
) -> str:
    return (
        f"ALTER TABLE {definition.database}.{definition.name} "
        f"MODIFY SETTING {key} = {_render_setting_value(value)};"
    )


def render_alter_reset_setting(definition: TableDefinition, key: str) -> str:
    return f"ALTER TABLE {definition.database}.{definition.name} RESET SETTING {key};"


def render_alter_modify_ttl(definition: TableDefinition, ttl: str | None) -> str:
    if ttl is None:
        return f"ALTER TABLE {definition.database}.{definition.name} REMOVE TTL;"
    return f"ALTER TABLE {definition.database}.{definition.name} MODIFY TTL {ttl};"

File: packages/core/src/sql.ts
import type {
  ColumnDefinition,
  DictionaryAttribute,
  DictionaryDefinition,
  MaterializedViewDefinition,
  MaterializedViewRefresh,
  ProjectionDefinition,
  SchemaDefinition,
  SkipIndexDefinition,
  TableDefinition,
  ViewDefinition,
} from './model.js'
import { renderCodec } from './codec.js'
import { isPlainColumnReference, normalizeKeyColumns } from './key-clause.js'
import { renderProjectionBody } from './projection.js'
import { renderTextIndexType } from './text-index.js'
import { assertValidDefinitions } from './validate.js'

function renderDefault(value: string | number | boolean): string {
  if (typeof value === 'string') {
    if (value.startsWith('fn:')) return value.slice(3)
    return `'${value.replace(/'/g, "''")}'`
  }
  return String(value)
}

function renderColumn(col: ColumnDefinition): string {
  let out = `\`${col.name}\` ${col.nullable ? `Nullable(${col.type})` : col.type}`
  if (col.default !== undefined) out += ` DEFAULT ${renderDefault(col.default)}`
  if (col.comment) out += ` COMMENT '${col.comment.replace(/'/g, "''")}'`
  if (col.codec) out += ` ${renderCodec(col.codec)}`
  return out
}

function renderKeyClauseColumns(columns: string[], columnNames: Set<string>): string {
  return normalizeKeyColumns(columns)
    .map((column) =>
      // Quote a token when it names a declared column (including names that
      // need quoting like `user-id`) or is a bare identifier. Only true
      // expressions (e.g. `toStartOfHour(ts)`) are emitted verbatim.
      columnNames.has(column) || isPlainColumnReference(column) ? `\`${column}\`` : column
    )
    .join(', ')
}

function renderIndexType(idx: SkipIndexDefinition): string {
  switch (idx.type) {
    case 'minmax':
      return 'minmax'
    case 'set':
      return `set(${idx.maxRows})`
    case 'bloom_filter':
      return idx.falsePositiveRate !== undefined
        ? `bloom_filter(${idx.falsePositiveRate})`
        : 'bloom_filter'
    case 'tokenbf_v1':
      return `tokenbf_v1(${idx.sizeBytes}, ${idx.hashFunctions}, ${idx.randomSeed})`
    case 'ngrambf_v1':
      return `ngrambf_v1(${idx.ngramSize}, ${idx.sizeBytes}, ${idx.hashFunctions}, ${idx.randomSeed})`
    case 'text':
      return renderTextIndexType(idx)
  }
}

function renderTableSQL(def: TableDefinition): string {
  const columns = def.columns.map(renderColumn)
  const indexes = (def.indexes ?? []).map(
    (idx) =>
      `INDEX \`${idx.name}\` (${idx.expression}) TYPE ${renderIndexType(idx)} GRANULARITY ${idx.granularity}`
  )
  const projections = (def.projections ?? []).map(
    (projection) => `PROJECTION \`${projection.name}\` ${renderProjectionBody(projection)}`
  )
  const body = [...columns, ...indexes, ...projections].join(',\n  ')

  const columnNames = new Set(def.columns.map((column) => column.name))
  const clauses: string[] = []
  if (def.partitionBy) clauses.push(`PARTITION BY ${def.partitionBy}`)
  clauses.push(`PRIMARY KEY (${renderKeyClauseColumns(def.primaryKey, columnNames)})`)
  clauses.push(`ORDER BY (${renderKeyClauseColumns(def.orderBy, columnNames)})`)
  if (def.uniqueKey && def.uniqueKey.length > 0) {
    clauses.push(`UNIQUE KEY (${renderKeyClauseColumns(def.uniqueKey, columnNames)})`)
  }
  if (def.ttl) clauses.push(`TTL ${def.ttl}`)
  if (def.settings && Object.keys(def.settings).length > 0) {
    clauses.push(
      `SETTINGS ${Object.entries(def.settings)
        .map(([k, v]) => `${k} = ${v}`)
        .join(', ')}`
    )
  }
  if (def.comment) clauses.push(`COMMENT '${def.comment.replace(/'/g, "''")}'`)

  return `CREATE TABLE IF NOT EXISTS ${def.database}.${def.name}\n(\n  ${body}\n) ENGINE = ${def.engine}\n${clauses.join('\n')};`
}

function renderViewSQL(def: ViewDefinition): string {
  return `CREATE VIEW IF NOT EXISTS ${def.database}.${def.name} AS\n${def.as};`
}

function renderRefreshSettings(settings: Record<string, string | number>): string {
  return Object.entries(settings)
    .map(([k, v]) => `${k} = ${typeof v === 'string' ? `'${v.replace(/'/g, "''")}'` : v}`)
    .join(', ')
}

function renderDependsOn(dependsOn: Array<{ database: string; name: string }>): string {
  return dependsOn.map((dep) => `${dep.database}.${dep.name}`).join(', ')
}

/**
 * Renders the REFRESH clause block for a refreshable materialized view.
 * Grammar order: REFRESH EVERY|AFTER <interval> [OFFSET <interval>] [RANDOMIZE FOR <interval>]
 *                [DEPENDS ON <list>] [SETTINGS <kv>] [APPEND]
 * Note: EMPTY belongs after TO in CREATE, so it's NOT included here.
 */
function renderRefreshClause(refresh: MaterializedViewRefresh): string {
  const parts: string[] = []
  if (refresh.every) parts.push(`REFRESH EVERY ${refresh.every}`)
  else if (refresh.after) parts.push(`REFRESH AFTER ${refresh.after}`)
  if (refresh.offset) parts.push(`OFFSET ${refresh.offset}`)
  if (refresh.randomize) parts.push(`RANDOMIZE FOR ${refresh.randomize}`)
  if (refresh.dependsOn && refresh.dependsOn.length > 0) {
    parts.push(`DEPENDS ON ${renderDependsOn(refresh.dependsOn)}`)
  }
  if (refresh.settings && Object.keys(refresh.settings).length > 0) {
    parts.push(`SETTINGS ${renderRefreshSettings(refresh.settings)}`)
  }
  if (refresh.append) parts.push('APPEND')
  return parts.join(' ')
}

function renderMaterializedViewSQL(def: MaterializedViewDefinition): string {
  const header = `CREATE MATERIALIZED VIEW IF NOT EXISTS ${def.database}.${def.name}`
  const refreshBlock = def.refresh ? `\n${renderRefreshClause(def.refresh)}` : ''
  const toClause = ` TO ${def.to.database}.${def.to.name}`
  const emptyClause = def.refresh?.empty ? ' EMPTY' : ''
  return `${header}${refreshBlock}${toClause}${emptyClause} AS\n${def.as};`
}

/**
 * Renders ALTER TABLE ... MODIFY REFRESH for a refresh-only change.
 * Per live validation: APPEND must be re-included if present, because ClickHouse
 * treats omitting APPEND as "remove APPEND" and rejects.
 */
export function renderAlterModifyRefresh(def: MaterializedViewDefinition): string {
  if (!def.refresh) {
    throw new Error(
      `Cannot render MODIFY REFRESH for ${def.database}.${def.name}: refresh is not set`
    )
  }
  return `ALTER TABLE ${def.database}.${def.name} MODIFY ${renderRefreshClause(def.refresh)};`
}

function renderDictionaryAttribute(attr: DictionaryAttribute): string {
  let out = `\`${attr.name}\` ${attr.type}`
  if (attr.expression !== undefined) out += ` EXPRESSION ${attr.expression}`
  else if (attr.default !== undefined) out += ` DEFAULT ${renderDefault(attr.default)}`
  if (attr.hierarchical) out += ' HIERARCHICAL'
  if (attr.bidirectional) out += ' BIDIRECTIONAL'
  if (attr.injective) out += ' INJECTIVE'
  if (attr.isObjectId) out += ' IS_OBJECT_ID'
  return out
}

function renderDictionaryRangeColumn(column: string, columnNames: Set<string>): string {
  return columnNames.has(column) || isPlainColumnReference(column) ? `\`${column}\`` : column
}

function renderDictionarySettings(settings: Record<string, string | number>): string {
  return Object.entries(settings)
    .map(([k, v]) => `${k} = ${typeof v === 'string' ? `'${v.replace(/'/g, "''")}'` : v}`)
    .join(', ')
}

export function renderDictionarySQL(def: DictionaryDefinition, replace = false): string {
  const verb = replace ? 'CREATE OR REPLACE DICTIONARY' : 'CREATE DICTIONARY IF NOT EXISTS'
  const attrs = def.attributes.map(renderDictionaryAttribute).join(',\n  ')
  const columnNames = new Set(def.attributes.map((a) => a.name))
  const pk = renderKeyClauseColumns(def.primaryKey, columnNames)
  const clauses = [
    `PRIMARY KEY ${pk}`,
    `SOURCE(${def.source})`,
    `LAYOUT(${def.layout})`,
    `LIFETIME(${def.lifetime})`,
  ]
  if (def.range) {
    clauses.push(
      `RANGE(MIN ${renderDictionaryRangeColumn(def.range.min, columnNames)} MAX ${renderDictionaryRangeColumn(def.range.max, columnNames)})`
    )
  }
  if (def.settings && Object.keys(def.settings).length > 0) {
    clauses.push(`SETTINGS(${renderDictionarySettings(def.settings)})`)
  }
  if (def.comment) clauses.push(`COMMENT '${def.comment.replace(/'/g, "''")}'`)
  return `${verb} ${def.database}.${def.name}\n(\n  ${attrs}\n)\n${clauses.join('\n')};`
}

export function toCreateSQL(def: SchemaDefinition): string {
  assertValidDefinitions([def])
  if (def.kind === 'table') return renderTableSQL(def)
  if (def.kind === 'view') return renderViewSQL(def)
  if (def.kind === 'dictionary') return renderDictionarySQL(def)
  return renderMaterializedViewSQL(def)
}

export function renderAlterAddColumn(def: TableDefinition, column: ColumnDefinition): string {
  return `ALTER TABLE ${def.database}.${def.name} ADD COLUMN IF NOT EXISTS ${renderColumn(column)};`
}

export function renderAlterModifyColumn(def: TableDefinition, column: ColumnDefinition): string {
  return `ALTER TABLE ${def.database}.${def.name} MODIFY COLUMN ${renderColumn(column)};`
}

export function renderAlterDropColumn(def: TableDefinition, columnName: string): string {
  return `ALTER TABLE ${def.database}.${def.name} DROP COLUMN IF EXISTS \`${columnName}\`;`
}

export function renderAlterRemoveCodec(def: TableDefinition, columnName: string): string {
  return `ALTER TABLE ${def.database}.${def.name} MODIFY COLUMN \`${columnName}\` REMOVE CODEC;`
}

export function renderAlterAddIndex(def: TableDefinition, index: SkipIndexDefinition): string {
  return `ALTER TABLE ${def.database}.${def.name} ADD INDEX IF NOT EXISTS \`${index.name}\` (${index.expression}) TYPE ${renderIndexType(index)} GRANULARITY ${index.granularity};`
}

export function renderAlterDropIndex(def: TableDefinition, indexName: string): string {
  return `ALTER TABLE ${def.database}.${def.name} DROP INDEX IF EXISTS \`${indexName}\`;`
}

export function renderAlterAddProjection(
  def: TableDefinition,
  projection: ProjectionDefinition
): string {
  return `ALTER TABLE ${def.database}.${def.name} ADD PROJECTION IF NOT EXISTS \`${projection.name}\` ${renderProjectionBody(projection)};`
}

export function renderAlterDropProjection(def: TableDefinition, projectionName: string): string {
  return `ALTER TABLE ${def.database}.${def.name} DROP PROJECTION IF EXISTS \`${projectionName}\`;`
}

export function renderAlterModifySetting(
  def: TableDefinition,
  key: string,
  value: string | number | boolean
): string {
  return `ALTER TABLE ${def.database}.${def.name} MODIFY SETTING ${key} = ${value};`
}

export function renderAlterResetSetting(def: TableDefinition, key: string): string {
  return `ALTER TABLE ${def.database}.${def.name} RESET SETTING ${key};`
}

export function renderAlterModifyTTL(def: TableDefinition, ttl: string | undefined): string {
  if (ttl === undefined) {
    return `ALTER TABLE ${def.database}.${def.name} REMOVE TTL;`
  }
  return `ALTER TABLE ${def.database}.${def.name} MODIFY TTL ${ttl};`
}
